### DISCUSSION TITLE
InvalidArgument: [pyAgrum] Invalid argument: Potential to copy has not the same domainSize.

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#13] [Q&A] [answered]
I am not sure why I am getting this error as the number of states is the same for the bn object and the data.

![Screenshot 2023-04-13 at 4 09 20 AM](https://user-images.githubusercontent.com/24616330/231695925-6d8868bb-abd6-40f4-927c-78315927d961.png)
[tr_data_graph.csv](https://github.com/agrumery/aGrUM/files/11219755/tr_data_graph.csv)

Below is how I got the error.

```
import pyAgrum as gum
sach_bn = gum.fastBN('PKA->Raf->Mek->Erk->Akt;'
                     'PKA->Mek;'
                     'PKA->Erk;'
                     'PKA->Akt;'
                     'PKA->Jnk;'
                     'PKA->P38;'
                    )

bn_without_S= gum.BayesNet(sach_bn)
learner = gum.BNLearner("tr_data_graph.csv")
learner.fitParameters(bn_without_S)
```

**Comment by phwuil** [comment]
You created a BN where all variables are binary. I guess this is not the case in your CSV.

![image](https://user-images.githubusercontent.com/5980636/231711438-d25ed193-dd67-4eba-9316-bed00d282ea9.png)

**Comment by kenneth-lee-ch** [comment]
But the data is indeed binary, that's why I got so confused...I also
attached the data file on the thread.

**Comment by phwuil** [comment]
![image](https://user-images.githubusercontent.com/5980636/231779233-5cce5fc9-ac53-4c84-b8a6-18f0089c4686.png)

PKA has only one value (always 0) ...  Maybe we should warn in this case ... or automatically add a second value ...

**Comment by kenneth-lee-ch** [comment]
I thought the fitparameter should just count the frequency based on the
given finite sample and look at the BN object for the exact domain size. In
this case, wouldn't it make sense to just return 0 for the entry in CPT of
PKA state 1?

Kenneth

**Reply by phwuil** [reply]
How can the BNLearner know how many modalities are not present in the database  ? Could PKA be equal to 0, 2, 3 ? 
So by default, the BNLearner takes the exact domain size of the data for the CPT it learns. So PKA is learned with only one value. And when `fitParameters` tries to fill a cpt of size `2*... `with a table of size `1*... `, it ends up with the error of domainSize...

However, there is a solution, you can bypass the automatic type induction of the BNLearner by explicitly giving it the exact type of each variable. You can add a BN as a parameter when constructing the BNLearner. This BN is not used for the structure but is used as a dictionary of all the variables and their types :

![image](https://user-images.githubusercontent.com/5980636/231808190-801e8dbc-a333-4154-adad-816a8b05793a.png)

**Reply by phwuil** [reply, marked as answer]
you can even pass the same bn :
```python
bn=gum.fastBN("Erk->Akt->PKA;Mek->Jnk->Raf->P38[5]") 
learner=gum.BNLearner("tr_data_graph.csv",bn) # use bn as template for the variables
learner.fitParameters(bn)
gnb.flow.row(bn.cpt("PKA"),bn.cpt("P38"))
```
